docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md

```diff
@@ -0,0 +1,456 @@
+doesn't need to explicitly call <xref:System.Diagnostics.Activity.Stop?displayProperty=nameWithType>.
+That simplifies the coding pattern.
+
+- <xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType> internally determines if
+there are any listeners recording the Activity. If there are no registered listeners or there are listeners which
+are not interested, StartActivity() will return `null` and avoid creating the Activity object. This
+is a performance optimization so that the code pattern can still be used in functions that are called very
+frequently.
+
+## Optional: Populate tags
+
+Activities support key-value data called Tags, commonly used to store any parameters of the work that
+may be useful for diagnostics. Update DoSomeWork() to include them:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
+            {
+                activity?.SetTag("foo", foo);
+                activity?.SetTag("bar", bar);
+                await StepOne();
+                await StepTwo();
+            }
+        }
+```
+
+```dotnetcli
+> dotnet run
+Activity.Id:          00-2b56072db8cb5a4496a4bfb69f46aa06-7bc4acda3b9cce4d-01
+Activity.DisplayName: SomeWork
+Activity.Kind:        Internal
+Activity.StartTime:   2021-03-18T10:37:31.4949570Z
+Activity.Duration:    00:00:01.5417719
+Activity.TagObjects:
+    foo: banana
+    bar: 8
+Resource associated with Activity:
+    service.name: MySample
+    service.instance.id: 25bbc1c3-2de5-48d9-9333-062377fea49c
+
+Example work done
+```
+
+#### Best Practices
+
+- As mentioned above, `activity` returned by <xref:System.Diagnostics.ActivitySource.StartActivity%2A?displayProperty=nameWithType>
+may be null. The null-coallescing operator ?. in C# is a convenient short-hand to only invoke
+<xref:System.Diagnostics.Activity.SetTag%2A?displayProperty=nameWithType> if activity is not null. The behavior is identical to
+writing:
+
+```csharp
+if(activity != null)
+{
+    activity.SetTag("foo", foo);
+}
+```
+
+- OpenTelemetry provides a set of recommended
+[conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions)
+for setting Tags on Activities that represent common types of application work.
+
+- If you are instrumenting functions with high performance requirements,
+<xref:System.Diagnostics.Activity.IsAllDataRequested?displayProperty=nameWithType> is a hint that indicates whether any
+of the code listening to Activities intends to read auxilliary information such as Tags. If no listener will read it then there
+is no need for the instrumented code to spend CPU cycles populating it. For simplicity this sample doesn't apply that
+optimization.
+
+## Optional: Adding Events
+
+Events are timestamped messages that can attach an arbitrary stream of additional diagnostic data to Activities. Add
```

Comment:
link to OpenTelemetry "Event" https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events

Reply:
I've thought about it and concluded I'd rather not lean too heavily on the OpenTelemetry specification at this point. I'm trying to strike a balance - on the one hand I want .NET's Activity instrumentation to stand on its own without needing an external definition, on the other I want developers to understand they can use OpenTelemetry to collect it. I am betting that developers familiar with OTel's model will find the mapping intuitive. For developer's not familiar with OTel's model I don't want them to feel that they have to understand OTel's spec to understand the API.